Move jackson-datatype-money module from zalando

sri-adarsh-kumar opened the pull request:
Moves jackson-datatype-money module from zalando

**Context**

Related to https://github.com/FasterXML/jackson-datatypes-misc/issues/5 and https://github.com/zalando/jackson-datatype-money/issues/224

From the above conversations, there is a consensus to move the [zalando/jackson-datatype-money](https://github.com/zalando/jackson-datatype-money) library as a sub-module of this repository.

In order to achieve this, I have moved the files almost 1:1 from the source repository.

Only major changes were related to tests. The source repo had Junit Jupiter tests with `@ParameterizedTest`. The destination library already had other tests using `junitParams.Parameters`, so this was adopted.

**Review Suggestions**

Please focus on 

- How to use correct License
- How to attribute credits correctly
- Package naming conventions

sri-adarsh-kumar committed:
Migrating Commit 7fd4c1e from zalando/jackson-datatype-money

on-behalf-of: Zalando OSS Community
Co-authored-by: Willi Schönborn <w.schoenborn@gmail.com>
Co-authored-by: Philipp Hirch <philipp.hirch@otto.de>
Co-authored-by: Jörn Horstmann <git@jhorstmann.net>
Co-authored-by: Lauri at Zalando <lauri.d.apple@gmail.com>
Co-authored-by: msparer <msp+git@molindo.at>
Co-authored-by: Alexander Yastrebov <alexander.yastrebov@zalando.de>
Co-authored-by: Alexey Venderov <alexey.venderov@zalando.de>
Co-authored-by: Alexander Yastrebov <yastrebov.alex@gmail.com>
Co-authored-by: Arnaud BOIVIN <arno94600@hotmail.com>
Co-authored-by: Bartosz Ocytko <bocytko@gmail.com>
Co-authored-by: Carlos Freund <carlosfreund@googlemail.com>
Co-authored-by: Dario Seidl <darioseidl@users.noreply.github.com>
Co-authored-by: Georgios Andrianakis <geoand@gmail.com>
Co-authored-by: Lauri at Zalando <lauri.apple@zalando.de>
Co-authored-by: Martin Prebio <martin.prebio@gmail.com>
Co-authored-by: Sean Sullivan <ssullivan@gilt.com>
Co-authored-by: Touko Vainio-Kaila <touko.vk@gmail.com>
Co-authored-by: lukasniemeier-zalando <lukas.niemeier@zalando.de>

sri-adarsh-kumar commented on javax-money/src/moditect/module-info.java:
```diff
@@ -0,0 +1,13 @@
+//TODO how is this generated
```
I am not sure how this file is generated.
Is there some documentation for this?

cowtowncoder replied:
I think it would be running

    ../mvnw moditect:generate-module-info

from new modules project dir. Let me try this on PR.

This provides the base, but must be modified; looking at other modules's `module-info.java` for inspiration.

cowtowncoder replied:
Hmmh. Alas, that command throws NPE.

sri-adarsh-kumar commented:
@cowtowncoder Please review when you have time.

cowtowncoder commented on javax-money/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,23 @@
+TODO What goes here? (This original one or another?)
```
If this code happened to be licensed under Apache License 2, we could just use modified version of other modules' `LICENSE`, like:

```
This copy of Jackson Jakarta JSONP extension module is licensed under the
Apache (Software) License, version 2.0 ("the License").
See the License for details about distribution rights, and the
specific rights regarding derivative works.

You may obtain a copy of the License at:

http://www.apache.org/licenses/LICENSE-2.0```

But as long as we know what the license used is (we can't change it), should modify above to refer to whatever license there is.

cowtowncoder replied:
That is: we have to use the original license -- unless we have original author(s) permission to use some other license.

cowtowncoder replied:
Ok so its MIT License instead. So... I guess we could just include full license like so, given it's copyright by Zalando. No changes needed.

But we do need to make sure `<license>` tag in `pom.xml` has MIT license (we might have parent pom defaulting to Apache 2)

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,124 @@
+    </description>
+    <url>https://github.com/FasterXML/jackson-datatypes-misc</url>
+    <licenses>
+        <license>
```
Assuming MIT License as per notes further down, this needs to be changed.

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,124 @@
+    <name>Jackson datatype: javax-money</name>
+    <packaging>jar</packaging>
+    <version>2.19.0-SNAPSHOT</version>
+    <description>Support for datatypes of Javax Money library (https://javamoney.github.io/)
```
Maybe also mention Money API spec is from JSR 354 (can link to https://javamoney.github.io/api.html I suppose)

cowtowncoder commented on javax-money/src/test/java/com/fasterxml/jackson/datatype/money/CurrencyUnitDeserializerTest.java:
```diff
@@ -0,0 +1,43 @@
+
+public final class CurrencyUnitDeserializerTest {
+
+    private final ObjectMapper unit = new ObjectMapper().findAndRegisterModules();
```
This is wrong: should specifically add just Money module and not whatever might happen to be in classpath. Ideally would use helper method in test base class or utility/helper class

cowtowncoder commented on javax-money/src/test/java/com/fasterxml/jackson/datatype/money/CurrencyUnitSchemaSerializerTest.java:
```diff
@@ -0,0 +1,25 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public final class CurrencyUnitSchemaSerializerTest {
```
We should remove this test: should NOT include cross-module dependency for tests because this now introduces a build/release dependency between modules.

cowtowncoder commented on javax-money/src/test/java/com/fasterxml/jackson/datatype/money/MonetaryAmountDeserializerTest.java:
```diff
@@ -0,0 +1,273 @@
+
+        final String content = "{\"amount\":29.95}";
+
+        final JsonProcessingException exception = assertThrows(
```
While technically correct, use of `JsonProcessingException` is not optimal as we should check specific subtype that is being thrown, JPE is the most generic Jackson exception base class.

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,124 @@
+
+        <dependency>
+            <groupId>org.javamoney.moneta</groupId>
+            <artifactId>moneta-core</artifactId>
```
Is this test-only dependency? Or does module require specific Money API implementation?

sri-adarsh-kumar replied:
The module supports Money API implementations like moneta's FastMoney, Money and RoundedMoney.

cowtowncoder replied:
Right: I was wondering if inclusion of specific  implementation was problematic wrt using something else -- but I guess one would just use Maven dependency exclude, or some other mechanism.

Or put another way: I can see such dependency necessary for testing but wasn't sure it was really needed as regular ("compile") dependency. If you think it is needed that's fine: just double-checking.

KeepItSimpleStupid replied:
Hi !
First, thanks for this `jackson-datatype-money` module to the Zalando team, we're using for a long time and are really pleased that it's about to be more integrated to the Jackson ecosystem :)

Specifically on this thread :
I would say that if the module is about to be named `javax-money`, maybe it's worth reconsidering the support of the deserialization of `org.javamoney.moneta.FastMoney`, `org.javamoney.moneta.Money`, `org.javamoney.moneta.RoundedMoney`, 3 concrete classes belonging to Moneta, the only implementation of javax-money spec and support only the deserialization of `javax.money.MonetaryAmount` ?
Except for tests, the usage of the package `org.javamoney` is only located in `MoneyModule.java` so it seems quite straightforward to remove the support and get rid of `moneta-core` as a compile time dependency.
For us who use only `javax.money.MonetaryAmount` in our code, it would ease the dependency management since we won't have to fear to break this module if we update Moneta.
Maybe that would mean an additional module `javax-money-moneta` if deserializing those 3 classes is still needed by some ? This module could depend and reuse some code from the `javax-money` though

cowtowncoder replied:
@KeepItSimpleStupid Sounds like a good idea to me; modules should focus on doing one concrete thing well and just that (as general guideline).

sri-adarsh-kumar replied:
I understand your viewpoint, especially with naming the module as javax but having a moneta part.

But as this migration work has already taken quite long, I would like to complete it before focusing on additional improvements.

Would that be ok?

cowtowncoder replied:
I think it's bit of a blocker tho, in that anything that gets published (as 2.19.0, say), will sort of need to be supported. Adding more support is much easier than removing existing functionality.

sri-adarsh-kumar replied:
I have removed the strict dependency on Moneta library.

This means, the JavaxMoneyModule can be used standalone for serializing MonetaryAmount types. 

Only if there is a need for deserializing, a suitable implementation of MonetaryAmount has to be provided.

Please check and let me know if the module matches your expectations.

I can then create a sibling module for the reference Moneta implementation.

cowtowncoder replied:
So, @KeepItSimpleStupid this works for you?

sri-adarsh-kumar replied:
From the thumbs up, I take that this is fine for KeepItSimpleStupid.

Following on this, a new module has been added to add support for Moneta implementations similar to before.

Please check if the naming, structure, etc are fine

KeepItSimpleStupid replied:
Yes, all fine for me, thank you !

sri-adarsh-kumar replied:
@cowtowncoder All review comments have been addressed.

We can merge if fine with you.

cowtowncoder replied:
@sri-adarsh-kumar I can go over this once more but it all sounds good for now.

One more thing: I know CLAs needed in general for future contributions as we discussed few weeks back. But for just this PR, I think that I would need CLA from you to cover initial merge. Is this something you could do? As I said, I don't need anything more at this point.

sri-adarsh-kumar replied:
According to [this comment](https://github.com/FasterXML/jackson-datatypes-misc/pull/48#issuecomment-2590423786), my work should be covered with the CCLA. 
Did I understand that correctly?

cowtowncoder replied:
If there was an CCLA, yes, but I haven't gotten one from Zalando. And based on discussions it sounded like there was hesitation to go with CLA. Either one is fine with me, but I do want this first inclusion to be covered by one or the other.

sri-adarsh-kumar replied:
Based on [this](https://github.com/FasterXML/jackson-datatypes-misc/pull/48#issuecomment-2590423786) comment, I think we will make a CCLA following the merge.

sri-adarsh-kumar replied:
@bocytko Is my assumption correct?

bocytko replied:
@cowtowncoder there is no hesitation on the CLA. Based on prior comments I thought that there is no dependency between the PR merge and CLA. All that matters is that abaf36d8c56f8bbd70429351268fe8b4d15b5100 and 487cbf95ed751a5d12d229024b81c14d1f8fb6da covering the OSS contribution do not require the CLA. Fine for the follow-on commits which accumulated in this PR :)

Time-wise, if bound to a printout+signature, I can get the CCLA signed and sent to you early next week.

cowtowncoder replied:
Ok I may have been unclear here. I think every PR should be done under a CLA or CCLA -- but for the initial one I'd only need one that cover the submitter (and not everyone who collaborated on source repository). And then from thereon the usual (C)CLA coverage for further PRs.

Timing is fine with me -- I will make sure this gets merge before 2.19.0 gets released.
I will be taking one week off in a week so there's no super hurry.

I hope this helps!

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/CurrencyUnitDeserializer.java:
```diff
@@ -0,0 +1,32 @@
+import static org.apiguardian.api.API.Status.MAINTAINED;
+
+@API(status = MAINTAINED)
+public final class CurrencyUnitDeserializer extends JsonDeserializer<CurrencyUnit> {
```
Usually should extend either `StdDeserializer` or (if applicable) `StdScalarDeserializer`.

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/CurrencyUnitSerializer.java:
```diff
@@ -0,0 +1,35 @@
+import static org.apiguardian.api.API.Status.MAINTAINED;
+
+@API(status = MAINTAINED)
+public final class CurrencyUnitSerializer extends StdSerializer<CurrencyUnit> {
```
Possibly should use `StdScalarSerializer`.

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,124 @@
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
```
As per my earlier note, cannot have this dependency -- adds unwanted cross-module dependency for builds, release process.

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,124 @@
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.kjetland</groupId>
```
This dependency OTOH is fine (as it's not part of Jackson module set -- and even if it had transitive dependencies to some Jackson components that'd be ok as it wouldn't complicate build/release process)

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MonetaryAmountDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public Object deserializeWithType(final JsonParser parser, final DeserializationContext context,
```
If extending `StdDeserializer` (or, `StdScaralDeserializer`), wouldn't need to implement this method, I think.

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MonetaryAmountDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+    private void checkPresent(final JsonParser parser, @Nullable final Object value, final String name)
+            throws JsonParseException {
+        if (value == null) {
+            throw new JsonParseException(parser, format("Missing property: '%s'", name));
```
Sub-optimal exception as it does not indicate value type that has the issue.
Should usually be passed `DeserializationContext` and call one of methods it provides for throwing more specific (semantic) exceptions.

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MonetaryAmountSerializer.java:
```diff
@@ -0,0 +1,122 @@
+    public void serializeWithType(final MonetaryAmount value, final JsonGenerator generator,
+            final SerializerProvider provider, final TypeSerializer serializer) throws IOException {
+
+        // effectively assuming no type information at all
```
... which is basically wrong :-(

(meaning, won't work with `@JsonTypeInfo`)

cowtowncoder replied:
No, I don't think it is blocker: good point on "let's move it first, then improve".

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MoneyModule.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
```
Should fix usage instead of suppressing deprecation warning...

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MoneyModule.java:
```diff
@@ -0,0 +1,171 @@
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public Version version() {
```
This is unnecessarily using `VersionUtil` instead of referencing generated `PackageVersion` and doing

```
public Version version() {
    return PackageVersion.VERSION;
}
```

cowtowncoder commented:
@sri-adarsh-kumar  Ok, so technically I think this looks ok (although I have suggestions and concerns wrt code actually).
But one question I have is whether we need someone else's approval for contribution -- or are you the author? (I understand you work for Zalando from profile?).

I think we need one or more CLAs (from https://github.com/FasterXML/jackson/ either individual or Corporate CLA):

1. From you (wrt directly adding files here), regardless of authorship of code
2. Additional CLA(s) from whoever can contribute code itself (if not you).

It might be simplest to get CCLA from Zalando, although that is not a requirement if authors can give individual CLAs.

sri-adarsh-kumar committed:
Respond to review comments

Changed License reference (WIP)
Removed cross-module dependency to jsonSchema
Testing specific Modules instead of findAndRegisterModules
Assert subtypes of JSONProcessingException
CurrencyUnitDeserializer extends StdScalarDeserializer
CurrencyUnitSerializer extends StdScalarSerializer
MonetaryAmountDeserializer throws semantic exceptions using DeserializationContext
MoneyModule version uses PackageVersion

sri-adarsh-kumar committed:
Move Money.md in from source repo

cowtowncoder commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/MoneyModule.java:
```diff
@@ -0,0 +1,168 @@
+
+
+@API(status = STABLE)
+public final class MoneyModule extends Module {
```
Should we actually rename this as `JavaxMoneyModule`?

... also, as a side-note, is there need for "Jakarta Money module"? (wrt javax licensing resulting in "forking" of javax APIs)

sri-adarsh-kumar replied:
Let me check on this.

cowtowncoder commented on javax-money/src/test/java/com/fasterxml/jackson/datatype/money/CurrencyUnitDeserializerTest.java:
```diff
@@ -0,0 +1,43 @@
+
+public final class CurrencyUnitDeserializerTest {
+
+    private final ObjectMapper unit = new ObjectMapper().registerModule(new MoneyModule());
```
One minor request (apologies for forgetting to mention initially): let's use

    JsonMapper.builder().addModule(new MoneyModule()).build();

the reason begin that ObjectMappers are immutable in Jackson 3.0, so merging this module is easier if we use builder pattern.

cowtowncoder commented:
@sri-adarsh-kumar Thank you for updates, good job so far!

The other thing then has to do with CLAs, permissions -- see my earlier note.
If we can get that through, I should be able to merge this pr!
(getting it merged into `master` for 3.0 will have its challenges but one thing at a time :) )

sri-adarsh-kumar committed:
Use hand-crafted moditect date and change ObjectMapper in test

sri-adarsh-kumar commented:
Thanks for all the support. 
I am working on the CLA process with my organisation.

cowtowncoder commented on javax-money/README.md:
````diff
@@ -0,0 +1,214 @@
+
+```java
+ObjectMapper mapper = new ObjectMapper()
+        .registerModule(new MoneyModule());
````
Let's change to use new-er "addModule":

```json
ObjectMapper mapper = JsonMapper.builder()
  .addModule(...)
  .build();
```

sri-adarsh-kumar replied:
Updated.

cowtowncoder commented on javax-money/README.md:
````diff
@@ -0,0 +1,214 @@
+
+```java
+ObjectMapper mapper = new ObjectMapper()
+        .registerModule(new MoneyModule().withQuotedDecimalNumbers());
````
same here (addModule())

sri-adarsh-kumar replied:
Updated.

cowtowncoder commented:
I think code is looking almost ready for merge, so just need to deal with CLA question and it could be merged!

sri-adarsh-kumar commented:
We are discussing the [CLA](https://github.com/FasterXML/jackson/blob/master/contributor-agreement-corporate.txt) internally.

Our source repository has some contributions made by non-Zalando employees, so we are wondering how this should be treated.

The contributors have licensed it to us (and everyone else) under the MIT license.

Section 7 of the CLA asks to submit the work separately but as this contribution is from multiple people, we are not sure how this should be handled.

Do you have any suggestions?

cowtowncoder commented:
Maybe marking license as MIT (in `pom.xml` and README.md, and `LICENSE` file for sub-module) would be enough. And getting CLA from just Zalando employees (or just individuals who plan to contribute) -- so CLA for future work -- would work?
Put another way, CLA would only needed for changes after inclusion (and from individuals handling addition); and keeping MIT license to cover initial contributions.

I definitely do not want you to have to try to chase down all contributors.

I am not a lawyer obviously but it seems to me the intent for licensing and CLAs would be compatible with this approach.

bocytko commented on javax-money/src/main/resources/META-INF/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2015-2016 Zalando SE
```
Are we updating the year to 2024 (or 2025)?

sri-adarsh-kumar replied:
Changed to 2025.

bocytko commented:
> Maybe marking license as MIT (in `pom.xml` and README.md, and `LICENSE` file for sub-module) would be enough. And getting CLA from just Zalando employees (or just individuals who plan to contribute) -- so CLA for future work -- would work? Put another way, CLA would only needed for changes after inclusion (and from individuals handling addition); and keeping MIT license to cover initial contributions.
> 
> I definitely do not want you to have to try to chase down all contributors.
> 
> I am not a lawyer obviously but it seems to me the intent for licensing and CLAs would be compatible with this approach.

Thank you for this proposal @cowtowncoder. We discussed it with our lawyers and got green light for: MIT + Corporate CLA for Zalando employee contributions following the inclusion.

@sri-adarsh-kumar will initiate any remaining steps, if required for the merge.

cowtowncoder commented:
@bocytko Sounds like we just need to ensure MIT LICENSE information for module (ideally as part of PR). And then CCLA, although that can follow merge and just pre-date further PRs.

sri-adarsh-kumar committed:
Rename MoneyModule, use addModule in ReadMe, update License year

cowtowncoder commented on javax-money/pom.xml:
```diff
@@ -0,0 +1,119 @@
+        <dependency>
+            <groupId>org.apiguardian</groupId>
+            <artifactId>apiguardian-api</artifactId>
+            <version>1.1.2</version>
```
Also, this... if these are annotations, should probably use scope of `provided`?

At least it would seem like this would not be typical `compile` dependency for actual functionality?

sri-adarsh-kumar committed:
Remove strict Moneta dependency, Update docs

sri-adarsh-kumar committed:
Add WIP Moneta Module

sri-adarsh-kumar committed:
Add WIP Moneta Module

cowtowncoder committed:
Merge branch '2.19' into 2.19

sri-adarsh-kumar committed:
Add WIP Moneta Module

sri-adarsh-kumar committed:
Merge branch '2.19' of https://github.com/sri-adarsh-kumar/jackson-datatypes-misc into 2.19

sri-adarsh-kumar committed:
Review docs

sri-adarsh-kumar committed:
Update parent README

KeepItSimpleStupid commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/JavaxMoneyModule.java:
```diff
@@ -0,0 +1,115 @@
+        deserializers.addDeserializer(CurrencyUnit.class, new CurrencyUnitDeserializer());
+        if (Objects.nonNull(this.amountFactory) && Objects.nonNull(this.implementationClass)) {
+            deserializers.addDeserializer(MonetaryAmount.class, new MonetaryAmountDeserializer<>(amountFactory, names));
+            // for reading into concrete implementation types
+            deserializers.addDeserializer(implementationClass, new MonetaryAmountDeserializer<>(amountFactory, names));
```
I believe we don't need those lines (et the `implementationClass` field so) ? If we need to register additional serializers, it will be donein the `MonetaMoneyModule` you just introduced ?

KeepItSimpleStupid replied:
But if a new implementation of Javax Money appears (I don't know any, but maybe you've one internal to Zalando ?), don't you think it should be treated like Moneta has been in this PR, with a dedicated Jackson module ?
If you preserve the `amountFactory` field, and build the JavaMoneyModule this way (assuming you hava a `MyCustomImplementation` implementing `MonetaryAmount`)
```java
new JavaMoneyModule().withMonetaryAmountFactory(MyCustomImplemention::of)
```
You won't be able to deserialize fields of type `MyCustomImplementation` directly, that's true, but you'll know that all your deserialized `MonetaryAmount` fields are of concrete type `MyCustomImplementation`. So with one cast you could get back your `MyCustomImplementation` if really needed. But isn't the point of a spec to mask these implementation details ?

KeepItSimpleStupid replied:
Thank you for taking the time to consider my comments, I'll do a last round of review and then it should be good for me, at least from a Javax Money user point of view !
I leave the Jackson considerations to @cowtowncoder ;)

KeepItSimpleStupid commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/JavaxMoneyModule.java:
```diff
@@ -0,0 +1,115 @@
+
+        final SimpleDeserializers deserializers = new SimpleDeserializers();
+        deserializers.addDeserializer(CurrencyUnit.class, new CurrencyUnitDeserializer());
+        if (Objects.nonNull(this.amountFactory) && Objects.nonNull(this.implementationClass)) {
+            deserializers.addDeserializer(MonetaryAmount.class, new MonetaryAmountDeserializer<>(amountFactory, names));
+            // for reading into concrete implementation types
+            deserializers.addDeserializer(implementationClass, new MonetaryAmountDeserializer<>(amountFactory, names));
+        }
```
I believe it could be replaced with the canonical way to build a new MonetaryAmount, i.e.

```java
javax.money.Monetary.getDefaultAmountFactory()
                    .setNumber(number)
                    .setCurrency(currency)
                    .create()
```

So something like
```suggestion
        deserializers.addDeserializer(MonetaryAmount.class, new MonetaryAmountDeserializer<>((amount, currency) -> Monetary.getDefaultAmountFactory()
                                                             .setNumber(number)
                                                             .setCurrency(currency)
                                                             .create(),
                                 names));
```

sri-adarsh-kumar replied:
Thanks for the pointer. I edited the method to use the SPI capabilities.

KeepItSimpleStupid replied:
I didn't see the `amountFactory` field at first but I now get its usage.
Maybe the constructor passing a null `amountFactory` should get a lambda based on `Monetary.getDefaultAmountFactory()...create()` instead ? This way you would unify the logic (no need to check if `amountFactory` is null or not

From this 
```java
public JavaxMoneyModule() {
        this(new DecimalAmountWriter(), FieldNames.defaults(), MonetaryAmountFormatFactory.NONE, null, null);
}
```
to this
```java
public JavaxMoneyModule() {
        this(new DecimalAmountWriter(), FieldNames.defaults(), MonetaryAmountFormatFactory.NONE, null, (amount, currency) -> Monetary.getAmountFactory(c)
                                                                                                                                     .setNumber(amount)
                                                                                                                                     .setCurrency(currency)
                                                                                                                                     .create());
}
```

sri-adarsh-kumar replied:
Thanks for the suggestion. I have refactored as mentioned.

KeepItSimpleStupid commented on javax-money-moneta/src/main/java/com/fasterxml/jackson/datatype/money/moneta/MonetaMoneyModule.java:
```diff
@@ -0,0 +1,160 @@
+        this.baseModule.setupModule(context);
+
+        final SimpleDeserializers deserializers = new SimpleDeserializers();
+        deserializers.addDeserializer(CurrencyUnit.class, new CurrencyUnitDeserializer());
+        deserializers.addDeserializer(MonetaryAmount.class, new MonetaryAmountDeserializer<>(amountFactory, names));
```
I believe those 2 serializers are already registered by the JavaxMoneyModule whose `setupModule` is called a few line above, so no need to duplicate them here ?

sri-adarsh-kumar replied:
You are correct. With the SPI change, this line has been removed.

KeepItSimpleStupid reviewed (commented):
A few comments, thank you :)

sri-adarsh-kumar committed:
Use SPI for MonetaryAmountFactory

sri-adarsh-kumar committed:
Deserialize MonetaryAmount with Amount Factory

sri-adarsh-kumar committed:
Migrate Javax Money and Moneta modules to Junit 5

KeepItSimpleStupid commented on javax-money-moneta/src/main/java/com/fasterxml/jackson/datatype/money/moneta/MonetaMoneyModule.java:
```diff
@@ -0,0 +1,148 @@
+public final class MonetaMoneyModule extends Module {
+
+    private final JavaxMoneyModule baseModule;
+    private final MonetaryAmountFactory<? extends MonetaryAmount> amountFactory;
```
I think this `amountFactory` field is never read so it can be removed ?

sri-adarsh-kumar replied:
After the above requested change, this will be needed to support the appropriate amount factory for MonetaryAmount.

This will be Money by default, or FastMoney or others if configured as such.

KeepItSimpleStupid reviewed (commented):
A few more comments

sri-adarsh-kumar committed:
Support for Moneta types moved from Javax to Moneta Module

sri-adarsh-kumar commented on javax-money/src/main/java/com/fasterxml/jackson/datatype/money/JavaxMoneyModule.java:
```diff
@@ -0,0 +1,115 @@
+        //Scan all registered MonetaryAmount types and add a default deserializer for them
+        for (Class c : Monetary.getAmountTypes()) {
+            deserializers.addDeserializer(c, new MonetaryAmountDeserializer<>((amount, currency) -> Monetary.getAmountFactory(c).setNumber(amount).setCurrency(currency).create(), names));
+        }
```
I have updated the code as requested.

KeepItSimpleStupid commented on javax-money-moneta/src/main/java/com/fasterxml/jackson/datatype/money/moneta/MonetaMoneyModule.java:
```diff
@@ -0,0 +1,167 @@
+        final SimpleDeserializers deserializers = new SimpleDeserializers();
+        deserializers.addDeserializer(CurrencyUnit.class, new CurrencyUnitDeserializer());
+
+        //Use provided amountFactory to deserialize a MonetaryAmount
+        deserializers.addDeserializer(MonetaryAmount.class, new MonetaryAmountDeserializer<>(amountFactory, names));
```
Thanks for the modifications :) 

If the default constructor becomes
```java
public MonetaMoneyModule() {
        this(new JavaxMoneyModule().withMonetaryAmountFactory(Money::of), FieldNames.defaults(),
                Money::of, FastMoney::of, Money::of, RoundedMoney::of);
    }
```
and the method `withMonetaryAmountFactory` becomes
```java
    private <T extends MonetaryAmount> MonetaMoneyModule withMonetaryAmountFactory(final MonetaryAmountFactory<T> amountFactory) {
        return new MonetaMoneyModule(baseModule.withMonetaryAmountFactory(amountFactory), names,
                fastMoneyFactory, moneyFactory, roundedMoneyFactory);
    }
```
I think you can get rid of both :
- this extra deserializer for `MonetaryAmount` and rely only on the one defined in the baseModule
- the `amountFactory` field of this `MonetaMoneyModule` class

sri-adarsh-kumar replied:
You are correct. Changed now.

I had previously assumed that the deserializing logic in MonetaModule will overwrite the Base Module
But this appears not to be the case.

KeepItSimpleStupid replied:
Thanks !
You can get rid of the extra dezerializer for CurrencyUnit as well ;) 
And I think it will all look good to me after that !

sri-adarsh-kumar replied:
D'oh, right again. Done.

sri-adarsh-kumar committed:
Remove Amount Factory from MonetaModule

sri-adarsh-kumar committed:
Remove CurrencyUnit deserializing from MonetaModule

cowtowncoder commented:
CCLA received, hope to merge this RSN.

cowtowncoder committed:
Merge branch '2.19' into 2.19

cowtowncoder commented on javax-money-moneta/README.md:
```diff
@@ -0,0 +1,198 @@
+# Jackson Datatype Money
```
Should probably be "Jackson Datatype Moneta"?

sri-adarsh-kumar replied:
Thanks. I have fixed it.

sri-adarsh-kumar committed:
Fix typo in Moneta Readme

cowtowncoder commented:
Hmmh. Ok, I did approve: but realized I do think module names should be tweak slightly.
Here's what I plan to change before merge:

1. Main javax.money module: change artifactId to `jackson-datatype-javax-money` (since we have `jackson-datatype-joda-money`) -- not just `jackson-datatype-money`
2. Moneta module: leave artifactId as `jackson-datatype-moneta`

cowtowncoder committed:
Bit of refactoring

cowtowncoder committed:
Fixing things wrt refactoring

cowtowncoder committed:
Merge branch '2.19' into 2.19

cowtowncoder committed:
minor reording of README

cowtowncoder committed:
Merge branch '2.19' of github.com:sri-adarsh-kumar/jackson-datatypes-misc into sri-adarsh-kumar/2.19

cowtowncoder committed:
Merge branch '2.19' into 2.19

cowtowncoder committed:
Minor to fix to `module-info.java`s

cowtowncoder committed:
Minor fixes to READMEs

cowtowncoder commented:
@sri-adarsh-kumar Ok things look good with one exception: trying to do #57 for Jackson 3 (branch `3.x`) and basically, as far as I know & with all the info I have, it looks like it may not be possible to support JPMS (module-info.java) with Lombok, as things are currently.

I was wondering if there is any chance to de-Lombokify `javax-money` and `moneta` modules?
It looks like number of classes that actually use it is quite small, and I think it is possible to use Lombok to just generate sources, check in?

Alternatively if someone did figure out a way to fix Maven build to work, that'd be great outcome too -- but based on googling, chatgpt'ing, I am bit pessimistic about that being doable.

sri-adarsh-kumar commented:
@cowtowncoder I see from https://github.com/FasterXML/jackson-datatypes-misc/pull/59 and https://github.com/FasterXML/jackson-datatypes-misc/pull/60 you have done this already.

Pls let me know if there are any open tasks for me.

cowtowncoder commented:
@sri-adarsh-kumar Yes, sorry, forgot to update here.

I think we are good -- thank you for the follow up!
